protocols/upnp/src/behaviour.rs

```diff
@@ -0,0 +1,581 @@
+                                        log::debug!(
+                                            "failed to map upnp mapped {} for {} protocol: {err}",
+                                            mapping.internal_addr,
+                                            mapping.protocol
+                                        );
+                                    }
+                                    _ => {
+                                        unreachable!()
+                                    }
+                                }
+                            }
+                            GatewayEvent::Removed(mapping) => {
+                                log::debug!(
+                                    "succcessfully removed UPnP mapping {} for {} protocol",
+                                    mapping.internal_addr,
+                                    mapping.protocol
+                                );
+                                self.mappings
+                                    .remove(&mapping)
+                                    .expect("mapping should exist");
+                            }
+                            GatewayEvent::RemovalFailure(mapping, err) => {
+                                log::debug!(
+                                    "could not remove UPnP mapping {} for {} protocol: {err}",
+                                    mapping.internal_addr,
+                                    mapping.protocol
+                                );
+                                if let Err(err) = gateway
+                                    .sender
+                                    .try_send(GatewayRequest::RemoveMapping(mapping.clone()))
+                                {
+                                    log::debug!(
+                                        "could not request port removal for {} on the gateway: {}",
+                                        mapping.multiaddr,
+                                        err
+                                    );
+                                }
+                            }
+                        }
+                    }
+
+                    // Renew expired and request inactive mappings.
+                    self.mappings.renew(&mut self.config, gateway, cx);
+                    return Poll::Pending;
+                }
+                GatewayState::GatewayNotFound => {
+                    return Poll::Ready(ToSwarm::GenerateEvent(Event::GatewayNotFound));
+                }
+            }
+        }
+    }
+}
+
+/// Extracts a [`SocketAddrV4`] and [`PortMappingProtocol`] from a given [`Multiaddr`].
+///
+/// Fails if the given [`Multiaddr`] does not begin with an IP
+/// protocol encapsulating a TCP or UDP port.
+fn multiaddr_to_socketaddr_protocol(
+    mut addr: Multiaddr,
+) -> Result<(SocketAddr, PortMappingProtocol), ()> {
+    let mut port = None;
+    let mut protocol = None;
+    while let Some(proto) = addr.pop() {
```

Comment:
Wouldn't reading the address from the front, e.g. via `addr.into_iter().next()` be easier to reason about? One can return early (`Err((())`) on e.g. `/dns4`, ...